tests/assembly/s390x-vector-abi.rs

```diff
@@ -0,0 +1,299 @@
+    x: &TransparentWrapper<i8x8>,
+) -> TransparentWrapper<i8x8> {
+    *x
+}
+// CHECK-LABEL: vector_transparent_wrapper_ret:
+// CHECK: vl %v24, 0(%r2), 3
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_transparent_wrapper_ret(
+    x: &TransparentWrapper<i8x16>,
+) -> TransparentWrapper<i8x16> {
+    *x
+}
+// CHECK-LABEL: vector_transparent_wrapper_ret_large:
+// z10: vl %v0, 16(%r3), 4
+// z10-NEXT: vl %v1, 0(%r3), 4
+// z10-NEXT: vst %v0, 16(%r2), 4
+// z10-NEXT: vst %v1, 0(%r2), 4
+// z10-NEXT: br %r14
+// z13: vl %v0, 0(%r3), 4
+// z13-NEXT: vl %v1, 16(%r3), 4
+// z13-NEXT: vst %v1, 16(%r2), 4
+// z13-NEXT: vst %v0, 0(%r2), 4
+// z13-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_transparent_wrapper_ret_large(
+    x: &TransparentWrapper<i8x32>,
+) -> TransparentWrapper<i8x32> {
+    *x
+}
+
+// CHECK-LABEL: vector_arg_small:
+// CHECK: vlgvg %r2, %v24, 0
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_arg_small(x: i8x8) -> i64 {
+    unsafe { *(&x as *const i8x8 as *const i64) }
+}
+// CHECK-LABEL: vector_arg:
+// CHECK: vlgvg %r2, %v24, 0
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_arg(x: i8x16) -> i64 {
+    unsafe { *(&x as *const i8x16 as *const i64) }
+}
+// CHECK-LABEL: vector_arg_large:
+// CHECK: lg %r2, 0(%r2)
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_arg_large(x: i8x32) -> i64 {
+    unsafe { *(&x as *const i8x32 as *const i64) }
+}
+
+// CHECK-LABEL: vector_wrapper_arg_small:
+// CHECK: vlgvg %r2, %v24, 0
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_wrapper_arg_small(x: Wrapper<i8x8>) -> i64 {
+    unsafe { *(&x as *const Wrapper<i8x8> as *const i64) }
+}
+// CHECK-LABEL: vector_wrapper_arg:
+// CHECK: vlgvg %r2, %v24, 0
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_wrapper_arg(x: Wrapper<i8x16>) -> i64 {
+    unsafe { *(&x as *const Wrapper<i8x16> as *const i64) }
+}
+// CHECK-LABEL: vector_wrapper_arg_large:
+// CHECK: lg %r2, 0(%r2)
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_wrapper_arg_large(x: Wrapper<i8x32>) -> i64 {
+    unsafe { *(&x as *const Wrapper<i8x32> as *const i64) }
+}
+
+// CHECK-LABEL: vector_wrapper_with_zst_arg_small:
+// CHECK: vlgvg %r2, %v24, 0
+// CHECK-NEXT: br %r14
+#[cfg_attr(no_vector, target_feature(enable = "vector"))]
+#[no_mangle]
+unsafe extern "C" fn vector_wrapper_with_zst_arg_small(x: WrapperWithZst<i8x8>) -> i64 {
```

Comment:
The current implementation of this PR ignores ZST (zero sized type), is this incorrect? Or is the current implementation where ZST is specifically ignored correct?

From the ABI doc mentioned in PR description:

> ‹vector_arg›: A vector_arg has one of the following types:
> - Any vector type whose size is 16 bytes or less.
> - A structure equivalent to such a vector type, where “equivalent” has the same meaning as for double_or_float.

And double_or_float is:

> ‹double_or_float›: A double_or_float is one of the following:
> - A float or _Decimal32.
> - A double or _Decimal64.
> - A structure equivalent to one of the above. A structure is equivalent to a type 𝑇 if and *only if it has exactly one member, which is either of type 𝑇 itself or a structure equivalent to type 𝑇*.

Looking at the existing code, is the handling of ZST platform dependent? (https://github.com/rust-lang/rust/pull/125854)

https://github.com/rust-lang/rust/blob/730ecfe194e6046bd03b94ba2f73937d6a9929d2/compiler/rustc_target/src/callconv/s390x.rs#L45-L54

Reply:
This is unfortunately a bit more complicated due to historical reasons.   For the purposes of deciding whether an aggregate is a "structure equivalent to a float or vector", zero-sized members *in C* are *not* ignored.  The presence of a zero-sized bitfield, nested struct, or nested array member does make the aggregate *not equivalent*, i.e. it will be passed in GPRs and not in FPRs/VRs.

However, for C++ types, there are some zero-sized members that *are* ignored; this applies in particular to zero-sized base classes as well as C++20 empty data members (marked with `[[no_unique_address]]`).  This is not currently documented in the ABI (but then again, the ABI doesn't document anything about C++ ...).

The question is, what does this now mean for Rust?   My understanding is that Rust here attempts to be ABI-compatible with *C* only (not C++), and therefore I think it makes sense that Rust should be treated like C, i.e. zero-sized members are not ignored.  Current Rust code seems to correctly implement this for FP types via `is_single_fp_element`, so I think this PR should do the same for vector types.

As to the second question, what if the whole argument is zero-sized?  Both the strict reading of the ABI document and the actual GCC/clang implementation agree that this is passed via implicit pointer (which arguably doesn't make much sense, but we cannot change this anymore).  So the current Rust code looks correct to me.